src/auth/downscopedclient.ts

```diff
@@ -0,0 +1,254 @@
+    // The Downscoped access token will be returned in GetAccessTokenResponse format.
+    // If cached access token is unavailable or expired, force refresh.
+    if (
+      !this.cachedDownscopedAccessToken ||
+      this.isExpired(this.cachedDownscopedAccessToken)
+    ) {
+      await this.refreshAccessTokenAsync();
+    }
+    // Return Downscoped access token in GetAccessTokenResponse format.
+    return {
+      token: this.cachedDownscopedAccessToken!.access_token,
+      res: this.cachedDownscopedAccessToken!.res,
+    };
+  }
+
+  /**
+   * The main authentication interface. It takes an optional url which when
+   * present is the endpoint> being accessed, and returns a Promise which
+   * resolves with authorization header fields.
+   *
+   * The result has the form:
+   * { Authorization: 'Bearer <access_token_value>' }
+   */
+  async getRequestHeaders(): Promise<Headers> {
+    throw new Error('Not implemented.');
+  }
+
+  /**
+   * Provides a request implementation with OAuth 2.0 flow. In cases of
+   * HTTP 401 and 403 responses, it automatically asks for a new access token
+   * and replays the unsuccessful request.
+   * @param opts Request options.
+   * @param callback callback.
+   * @return A promise that resolves with the HTTP response when no callback
+   * is provided.
+   */
+  request<T>(opts: GaxiosOptions): GaxiosPromise<T>;
+  request<T>(opts: GaxiosOptions, callback: BodyResponseCallback<T>): void;
+  request<T>(
+    opts: GaxiosOptions,
+    callback?: BodyResponseCallback<T>
+  ): GaxiosPromise<T> | void {
+    throw new Error('Not implemented.');
+  }
+
+  /**
+   * Forces token refresh, even if unexpired tokens are currently cached.
+   * GCP access tokens are retrieved from authclient object/source credential.
+   * Thenm GCP access tokens are exchanged for downscoped access tokens via the
+   * token exchange endpoint.
+   * @return A promise that resolves with the fresh downscoped access token.
+   */
+  protected async refreshAccessTokenAsync(): Promise<CredentialsWithResponse> {
+    // Retrieve GCP access token from source credential.
+    const subjectToken = await (await this.authClient.getAccessToken()).token;
+    // Construct the STS credentials options.
+    const stsCredentialsOptions: sts.StsCredentialsOptions = {
+      grantType: STS_GRANT_TYPE,
+      requestedTokenType: STS_REQUEST_TOKEN_TYPE,
+      subjectToken: subjectToken as string,
+      subjectTokenType: STS_SUBJECT_TOKEN_TYPE,
+    };
+
+    // Exchange the source access token for a Downscoped access token.
+    const stsResponse = await this.stsCredential.exchangeToken(
+      stsCredentialsOptions,
+      undefined,
+      this.credentialAccessBoundary
+    );
+
+    // Save response in cached access token.
+    this.cachedDownscopedAccessToken = {
+      access_token: stsResponse.access_token,
+      expiry_date: new Date().getTime() + stsResponse.expires_in * 1000,
```

Comment:
Note, `expires_in` may not always be available. In that case you can use the authClient expiration time.

Reply:
that part is in another PR I sent yesterday:
https://github.com/googleapis/google-auth-library-nodejs/pull/1216